foqus_lib/framework/sdoe/nusf_dask.py

```diff
@@ -0,0 +1,204 @@
+# FOQUS Copyright (c) 2012 - 2023, by the software owners: Oak Ridge Institute
+# for Science and Education (ORISE), TRIAD National Security, LLC., Lawrence
+# Livermore National Security, LLC., The Regents of the University of
+# California, through Lawrence Berkeley National Laboratory, Battelle Memorial
+# Institute, Pacific Northwest Division through Pacific Northwest National
+# Laboratory, Carnegie Mellon University, West Virginia University, Boston
+# University, the Trustees of Princeton University, The University of Texas at
+# Austin, URS Energy & Construction, Inc., et al.  All rights reserved.
+#
+# Please see the file LICENSE.md for full copyright and license information,
+# respectively. This file is also available online at the URL
+# "https://github.com/CCSI-Toolset/FOQUS".
+#################################################################################
+import time
+from typing import Dict, List, Optional, Tuple, TypedDict, Union
+
+import dask.bag as db
+import numpy as np
+import pandas as pd  # only used for the final output of criterion
+from scipy.stats import rankdata
+
+from .distance import compute_dist, compute_min_params
+from .nusf import compute_dmat, scale_xs, scale_y, update_min_dist
+
+
+def criterion(
+    cand: pd.DataFrame,  # candidates
+    args: TypedDict(
+        "args",
+        {
+            "icol": str,
+            "xcols": List,
+            "wcol": str,
+            "max_iterations": int,
+            "mwr_values": List,
+            "scale_method": str,
+        },
+    ),  # maximum number of iterations, mwr values, scale method, index types
+    nr: int,  # number of restarts (each restart uses a random set of <nd> points)
+    nd: int,  # design size <= len(candidates)
+    mode: str = "maximin",
+    hist: Optional[pd.DataFrame] = None,
+    rand_seed: Union[int, None] = None,
+) -> Dict:
+
+    ncand = len(cand)
+    if hist is not None:
+        nhist = len(hist)
+    assert nd <= ncand, "Number of designs exceeds number of available candidates."
+
+    mode = mode.lower()
+    assert (
+        mode == "maximin"
+    ), "MODE {} not recognized for NUSF. Only MAXIMIN is currently supported.".format(
+        mode
+    )
+
+    T = args["max_iterations"]
+    mwr_vals = args["mwr_values"]
+    scale_method = args["scale_method"]
+
+    # index types
+    idx_np = [cand.columns.get_loc(i) for i in args["xcols"]]
+    idw_np = cand.columns.get_loc(args["wcol"])
+
+    cand_np_ = cand.to_numpy()
+    cand_np_unscaled = cand_np_.copy()
+
+    # Combine candidates and history before scaling
+    if hist is None:
+        cand_np_, _xmin, _xmax = scale_xs(cand_np_, idx_np)
+    else:
+        cand_np_, _xmin, _xmax = scale_xs(
+            np.concatenate((cand_np_, hist.to_numpy())), idx_np
+        )
+
+    rand_gen = np.random.default_rng(rand_seed)
+
+    def step(mwr_tuple: Tuple[int, [int], np.ndarray, Union[np.ndarray, None]]) -> Dict:
```

Comment:
I think the `[int]` inside the `Tuple` is what's causing the current errors, not sure if it's a typo (or rather... a `type`-o?... :drum

Reply:
yeah - that works in 3.11+ but not earlier - added fix